### ISSUE TITLE
Nasal task display / Nasal "top"

### ISSUE THREAD

**Issue by pedromorgan** [pedromorgan/flightgear-issues-test#563]
_From [stefan.g...@gmail.com](https://code.google.com/u/106730692913685497319/) on January 04, 2012 14:47:24_

As described in issue #558 , sometimes Nasal script go wild and eat up 100&#37; CPU power. For debugging and also for streamlining Nasal code, an additional function to the system monitor which provides an output of tasks' resource usage, similar to the output of the linux/unix tool top would be extremely helpful.

_Original issue: http://code.google.com/p/flightgear-bugs/issues/detail?id=564_

**Comment by pedromorgan** [comment]
_From [wonco...@googlemail.com](https://code.google.com/u/107832458638100707460/) on January 04, 2012 08:26:32_

I agree, as mentioned in \<a href=" https://code.google.com/p/flightgear-bugs/issues/detail?id=558#c3">comment 3 in `#558`</a>, I tried to look into this. 
But part of the problem seems to be that this info is currently not available in Nasal?

If I am not mistaken, Nasal is sort of agnostic to the namespace/function running after the codegen has finished. The best you could get out of the VM seems to be filename and line number (that would already be very helpful though).

Basically, you would need to create a time stamp for each Nasal timer/event and update the timer for each invocation. That would at least give you the file names and line numbers of the most resource hungry scripts.

**Comment by pedromorgan** [comment]
_From [wonco...@googlemail.com](https://code.google.com/u/107832458638100707460/) on February 12, 2012 07:08:56_

I still believe that it would be good to get in-depth information on the Nasal scripts that are running and their overall overhead.

After looking into the C++ source code: http://gitorious.org/fg/flightgear/blobs/next/src/Scripting/NasalSys.cxx It is my understanding that the handleTimer() method is responsible to call the registered Nasal code once the timer has expired: http://gitorious.org/fg/flightgear/blobs/next/src/Scripting/NasalSys.cxx#line1180 So, wrapping call() in between two SGTimeStamps should give us the execution time for the Nasal callback being invoked, right?

Next, we need to identify the timer being called (ideally, the file/and namespace.function).
But, the only useful info I have been able to find so far is using the file name and line number, like I mentioned earlier.

That would give us the call overhead for each invoked timer, which could be published into the property tree, so that a dialog could show a list of the most "expensive" timers that are run, by i.e. in the form of: foo.nas@271 

The same approach could probably be used for listeners that are regularly fired.

Any thoughts/comments?

Thanks

PS: Could someone please add a "nasal" tag to this?

**Comment by pedromorgan** [comment]
_From [pedromor...@gmail.com](https://code.google.com/u/112081649209717344658/) on December 28, 2012 08:10:16_

Added nasal tag..

Is this issue still valid.. ??

**Labels:** Nasal

**Comment by pedromorgan** [comment]
_From [wonco...@googlemail.com](https://code.google.com/u/107832458638100707460/) on December 28, 2012 08:19:49_

it definitely hasn't been implemented yet. And I'm not sure if anybody is working on this. I never really got any feedback in response to the questions and suggestions that I mentioned. And some of my patches have taken months to be reviewed, so I'm frankly not overly encouraged.
With some more feedback I might be convinced to look into this - assuming that there's someone who feels responsible/qualified to review a corresponding patch.
